File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/fragments/landing/LandingPresenter.java
package io.muun.apollo.presentation.ui.fragments.landing;

import io.muun.apollo.domain.action.fcm.ForceFetchFcmAction;
import io.muun.apollo.presentation.analytics.AnalyticsEvent;
import io.muun.apollo.presentation.ui.base.SingleFragmentPresenter;
import io.muun.apollo.presentation.ui.base.SingleFragmentView;
import io.muun.apollo.presentation.ui.base.di.PerFragment;
import io.muun.apollo.presentation.ui.signup.SignupPresenter;

import javax.inject.Inject;

@PerFragment
public class LandingPresenter extends SingleFragmentPresenter<SingleFragmentView, SignupPresenter> {

    private final ForceFetchFcmAction forceFetchFcm;

    @Inject
    public LandingPresenter(ForceFetchFcmAction forceFetchFcm) {
        this.forceFetchFcm = forceFetchFcm;
    }

    @Override
    public void afterSetUp() {
        super.afterSetUp();
        getParentPresenter().resumeSignupIfStarted();
        assertGooglePlayServicesPresent();

        // "Fire and forget" attempt to retrieve FCM token so its already available at signup/signin
        forceFetchFcm.run();
    }

    public void startSignup() {
        getParentPresenter().startSignup();
    }

    public void startLogin() {
        getParentPresenter().startLogin();
    }

    @Override
    protected AnalyticsEvent getEntryEvent() {
        return new AnalyticsEvent.S_GET_STARTED();
    }
}

File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/fragments/landing/LandingPresenter.kt
package io.muun.apollo.presentation.ui.fragments.landing

import io.muun.apollo.domain.action.fcm.ForceFetchFcmAction
import io.muun.apollo.domain.analytics.AnalyticsEvent
import io.muun.apollo.domain.analytics.AnalyticsEvent.S_GET_STARTED
import io.muun.apollo.presentation.ui.base.SingleFragmentPresenter
import io.muun.apollo.presentation.ui.base.SingleFragmentView
import io.muun.apollo.presentation.ui.base.di.PerFragment
import io.muun.apollo.presentation.ui.signup.SignupPresenter
import javax.inject.Inject

@PerFragment
open class LandingPresenter @Inject constructor(private val forceFetchFcm: ForceFetchFcmAction) :
    SingleFragmentPresenter<SingleFragmentView, SignupPresenter>() {

    override fun onSetUpFinished() {
        super.onSetUpFinished()
        parentPresenter.resumeSignupIfStarted()
        assertGooglePlayServicesPresent()

        // "Fire and forget" attempt to retrieve FCM token so its already available at wallet
        // creation or recovery.
        forceFetchFcm.run()
    }

    /**
     * Start wallet creation flow.
     */
    fun startCreateWallet() {
        parentPresenter.startSignup()
    }

    /**
     * Start wallet recovery flow.
     */
    fun startRecoverWallet() {
        parentPresenter.startLogin()
    }

    override fun getEntryEvent(): AnalyticsEvent? {
        return S_GET_STARTED()
    }
}